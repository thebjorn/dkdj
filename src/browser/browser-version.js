import $ from 'jquery';  // use dk.$ ?
import dkglobal from "../lifecycle/dkglobal";


if (typeof window === "undefined") {
    // running under node.js
    window = {
        navigator: {
            userAgent: ''
        }
    };
}


export default (function () {
    // $.browser disappeared in jQuery 1.9
    if ($.browser) return $.browser;
    if (dkglobal._browser) return this._browser;

    // from jquery.dynatree.js with modifications
    var matched, browser;

    var uaMatch = function( ua ) {
        ua = ua.toLowerCase();

        var match = /(opr)[\/]([\w.]+)/.exec( ua ) ||
                /(chrome)[ \/]([\w.]+)/.exec( ua ) ||
                /(version)[ \/]([\w.]+).*(safari)[ \/]([\w.]+)/.exec(ua) ||
                /(webkit)[ \/]([\w.]+)/.exec( ua ) ||
                /(opera)(?:.*version|)[ \/]([\w.]+)/.exec( ua ) ||
                /(msie) ([\w.]+)/.exec( ua ) ||
                ua.indexOf("trident") >= 0 && /(rv)(?::| )([\w.]+)/.exec( ua ) ||
                ua.indexOf("compatible") < 0 && /(mozilla)(?:.*? rv:([\w.]+)|)/.exec( ua ) ||
                [];

        var platform_match = /(ipad)/.exec( ua ) ||
                /(iphone)/.exec( ua ) ||
                /(android)/.exec( ua ) ||
                /(windows phone)/.exec(ua) ||
                /(win)/.exec( ua ) ||
                /(mac)/.exec( ua ) ||
                /(linux)/.exec( ua ) ||
                [];

        return {
            browser: match[ 3 ] || match[ 1 ] || "",
            version: match[ 2 ] || "0",
            platform: platform_match[0] || ""
        };
    };
    matched = uaMatch(window.navigator.userAgent);
    browser = {};
    if (matched.browser) {
        browser[matched.browser] = true;
        browser.version = matched.version;
        browser.versionNumber = parseInt(matched.version, 10);
    }
    if (matched.platform) browser[matched.platform] = true;
    if (browser.chrome || browser.opr || browser.safari) browser.webkit = true;
    if (browser.rv) {
        matched.browser = 'msie';
        browser.msie = true;
    }
    if (browser.opr) {
        matched.browser = 'opera';
        browser.opera = true;
    }
    if (browser.safari && browser.android) {
        matched.browser = 'android';
        browser.android = true;
    }
    browser.name = matched.browser;
    browser.platform = matched.platform;
    return browser;
}());
