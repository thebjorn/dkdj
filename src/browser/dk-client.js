
import dk from "../dk-obj";
import cookie from "./dk-cookie";
import jason from "../data/datacore/dk-json";


/*
 *  A csrf-aware ajax call.
 */
function _make_global_ajax_params(params) {
    const self = this;
    const csrf_safe = function (method) {
        return (/^(GET|HEAD|OPTIONS|TRACE)$/.test(method));
    };
    const beforeSend = params.beforeSend;
    params.beforeSend = function (xhr, settings) {
        xhr.setRequestHeader("X-dkjs", dk.version);
        if (!csrf_safe(params.type)) {
            xhr.setRequestHeader("X-CSRFToken", cookie.get('csrftoken'));
        }
        if (beforeSend) {
            beforeSend.call(self, xhr, settings);
        }
    };

    if (!csrf_safe(params.type)) {
        //params.crossDomain = false;
    }
    if (params.data && typeof params.data !== 'string') {
        params.data = JSON.stringify(params.data);
        params.dataType = 'json';
        params.contentType = 'application/json; charset=utf-8';
    }
    return params;    
}
export function ajax(params) {
    return dk.$.ajax(_make_global_ajax_params(params));
}
export async function async_ajax(params) {
    return await dk.$.ajax(_make_global_ajax_params(params));
} 

/*
 *  A better behaved ajax.
 *
 */
function _make_global_json_params(params) {
    params = dk.$.extend({
        cache: false,
        dataType: 'json',
        statusCode: {
            404: function () { alert("Page not found: " + params.url); },
            409: function (xhr, textStatus, error) {  // rate limiting error
                dk.log(409, xhr, textStatus, error);
                //alert(error);
            },
            500: function () {
                // display the server error to the user.
                if (!params.stop_500_redirect) {
                    document.location = params.url;
                }
            }
        },
        error: function (req, status, err) {
            dk.log("ERROR", req, status, err);
            document.location = params.url;
            //throw {error: "fetch-error", message: status + ' ' + err};
        },
        converters: {
            "text json": jason.parse
        },
        success: function () {
            // dk.log("SUCCESS:", data);
        }
    }, params);
    if (params.data) {
        params.type = 'POST';
    }
    return params;
}
export function json(params) {
    return ajax(_make_global_json_params(params));
}
export async function async_json(params) {
    return await async_ajax(_make_global_json_params(params));
}
