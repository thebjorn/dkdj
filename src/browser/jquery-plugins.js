

export function jq_links2popup(dk) {

    // create popup windows from links with class=popup
    dk.$('a.popup').click(function () {
        const w = window.open(
            dk.$(this).prop('href'),
            dk.$(this).prop('id') || "WindowName",
            ('width=' + dk.$(this).attr('width') +
            ',height=' + dk.$(this).attr('height') +
            ',resizeable=yes,' +
            'scrollbars=yes,' +
            'toolbar=no,' +
            'location=no,' +
            'directories=no,' +
            'status=no,' +
            'menubar=no,' +
            'copyhistory=no')
        );
        if (w) w.focus();
        return false;
    });

}



export function jq_help_button(dk) {
    const _helpfn = function (item) {
        dk.$(item).find('.help > .help-text').each(function () {
            let $help = dk.$(this).parent();
            let helptext = dk.$(this);
            // $help.css('position', 'relative');
            let trigger = dk.$('<div class="help-trigger"><a><dk-icon value="question-circle"/></a></div>');
            $help.prepend(trigger);
            $help.find('.help-trigger > a').addClass('xbtn').attr('tabindex', 10).popover({
                html: true,
                title: helptext.attr('title') || 'Hjelp..',
                container: 'body',
                trigger: 'focus',
                toggle: 'popover',
                placement: helptext.attr('placement') || 'auto left',
                content: helptext.html()
            });
        });
    };


    /*
        To get a little blue questionmark that displays help text when hoverered over,
        use the following markup:

            <div class="help">

                <div class="help-text" title="my title" hidden>
                    Help text goes here..
                </div>

                content that should get the blue questionmark goes here..

            </div>

        override .help>.help-trigger to change the palcement of the ?-mark.
        The defaults are in dkjs/css/dk-help.less
    */
    dk.$(document).ready(function () {
        _helpfn('body');
    });
}


export function jq_get_attributes(dk) {
    /*
     *  Return all attributes of a html object.
     *
     *  Usage::
     *
     *     $("<input type='text' value='5'>").getAttributes()
     *
     *          => {type: 'text', value: '5'}
     *
     *     $("<div data-foo="FOO" data-bar="BAR" pos=0).getAttributes()
     *
     *          => {
     *              data: {
     *                  foo: 'FOO',
     *                  bar: 'BAR'
     *              },
     *              pos: 0
     *          }
     *
     *  It is recommended that you use the html5 standard "data-"
     *  prefix for data attributes, although this is not required.
     *
     */
    dk.$.fn.getAttributes = function () {   // TODO: add prefix selector (e.g. all data- attributes).
        const elem = this,
            attr = {};

        const set_attr = function (res, name_parts, val) {
            //dk.debug('set_attr', res, name_parts, val);
            if (name_parts.length === 1) {
                res[name_parts[0]] = val;
                return;
            } else {
                const first = name_parts.shift();
                if (!res[first]) {
                    res[first] = {};
                }
                set_attr(res[first], name_parts, val);
            }
        };

        if (elem.length) dk.$.each(elem.get(0).attributes, function (v, n) {
            n = n.nodeName || n.name;
            v = elem.attr(n); // rely on $.fn.attr, it makes some filtering and checks
            if (v !== undefined && v !== false) set_attr(attr, n.split('-'), v);
        });

        return attr;
    };
}


export function jq_toggle_busy(dk) {
    /*
     *  Add, or remove, a semi-transparent shim above the selected item(s).
     *
     *  Usage::
     *
     *     this.widget().toggle_busy(true);
     *     this.widget().toggle_busy(false);
     */
    dk.$.fn.toggle_busy = function (state) {
        state = !!state;

        return this.each(function () {
            const $this = dk.$(this);
            const shimdata = $this.data('busy-shim');
            if (!state || shimdata) {
                dk.$('#' + shimdata).remove();
            } else if (state || !shimdata) {
                const height = $this.outerHeight();
                const width = $this.outerWidth();
                const offset = $this.offset();

                const shim = dk.$('<div/>').prop('id', 'busy-shim-' + dk.counter()).css({
                    position: 'absolute',
                    width: width,
                    height: height,
                    left: offset.left,
                    top: offset.top,
                    zIndex: 999,
                    backgroundColor: 'rgba(222,222,222,.5)'
                });
                dk.$('body').append(shim);
                $this.data('busy-shim', shim.prop('id'));
            }
        });
    };
}


export function jq_access_keys(dk) {
    /*
     *   $().accesskeys();
     */
    function possibilities(label, used) {
        const res = [];
        for (let i = 0; i < label.length; i++) {
            const ch = label[i];
            if (!used[ch]) {
                const t = {};
                t[ch] = label;
                res.push(Object.assign(t, used));
            }
        }
        return res;
    }

    function find_keys(labels, used) {
        if (labels.length === 1) {
            return possibilities(labels[0], used);
        } else {
            const _used_list = find_keys(labels.slice(-1), used);
            for (let i = 0; i < _used_list.length; i++) {
                const _used = _used_list[i];
                const _found_list = find_keys(labels.slice(0, -1), _used);
                if (_found_list.length > 0) return [_found_list[0]];
            }
            return [];
        }
    }

    function findKeys(labels, used) {
        const r = find_keys(labels, used || {});
        return r.length > 0 ? r[0] : null;
    }

    function wrap_access_key($item, key, settings) {
        const text = $item.text();
        if (text === "") return;
        const txt = text.toLowerCase();
        const pos = txt.indexOf(key);
        const isupper = text[pos] === text[pos].toUpperCase();
        if (isupper) key = key.toUpperCase();
        $item.html(text.slice(0, pos) + settings.wrap[0] + key + settings.wrap[1] + text.slice(pos + 1));
    }

    dk.$.fn.accesskeys = function (settings) {
        settings = dk.$.extend({
            skip: /[^\wæøå]/gi,                             // charachters to skip (invalid as access keys)
            keep: '[accesskey][accesskey!="-"]',            // selector of items to NOT modify
            modify: '[accesskey="-"]',                      // selector of items to modify
            wrap: ['<span class="accesskey">', '</span>']   // tag to wrap the access key with (null => don't wrap).
        }, settings);

        const used_chars = Array.from(dk.$(settings.keep).map(function (v, i) {
            return dk.$(v).attr('accesskey');
        }));
        const used = dk.zip_object(used_chars, dk.times(used_chars.length, () => '-'));

        const items = dk.$(settings.modify);
        const labels = Array.from(items.map(function (val, i) {
            return dk.$(val).text().trim().toLowerCase().replace(settings.skip, '');
        }));
        const item_map = dk.zip_object(labels, items);
        const akeys = findKeys(labels, used);
        if (akeys === null) {
            dk.error("Could not find valid combination of access keys.");
            return;
        }

        Object.keys(akeys).forEach(function (key) {
            const label = akeys[key];
            const $item = dk.$(item_map[label]);
            $item.attr('accesskey', key);
            if (settings.wrap !== null) wrap_access_key($item, key, settings);
        });

        return this;
    };
}
