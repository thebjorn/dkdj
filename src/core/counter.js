// LVL:0

let _counters = {
    _default: 1
};

/**
 *  Return a unique counter value.
 *  The optional 'name' parameter, will create a label
 *  'name'<counter>, where the counter starts from startval
 *  (default 1).
 *
 * @param {string}  name - the name of the counter.
 * @param {int}     startval - the starting value (default=1).
 * @returns {string}
 */
function counter(name, startval=0) {
    if (!name && !startval) return _counters._default++;
    if (!_counters[name]) {
        if (!startval) startval = 1;
        _counters[name] = startval;
    }
    return name + _counters[name]++;
}

export function _reset_counters() {
    _counters = {_default: 1};
}

export default counter;
