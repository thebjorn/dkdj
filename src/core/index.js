
var jqp = require('./dk-jqplugins.js');


/*
 *  Return a unique counter value.
 *  The optional ``name`` parameter, will create a label
 *  ``name``<counter>, where the counter starts from startval
 *  (default 1).
 */
const _counters = {
    _default: 1
};

export function counter(name, startval) {
    if (!name && !startval) return _counters._default++;
    if (!_counters[name]) {
        if (!startval) startval = 1;
        _counters[name] = startval;
    }
    return name + _counters[name]++;
}


/**
 * Convert ``id`` to a valid javascript identifier.
 *
 * @param id
 * @returns {*}
 */
export function id2name(id) {
    return id.replace(/-/g, '_');
}

/**
 * Convert `MyClass` -> my-class.
 *
 * @param clsname
 * @returns {string}
 */
export function cls2id(clsname) {
    return clsname.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
}


/**
 *  Count the occurences of letter in string `s`.
 * @param s
 * @param letter
 * @returns {*}
 */
export function count_char(s, letter) {
    return s.match(new RegExp(letter, 'g')).length;
}

/**
 * Similar to Python's textwrap.dedent(), which removes common/leading
 *  whitespace from the left of each line in txt.
 *
 * @param txt
 * @returns {*}
 */
export function dedent(txt) {
    if (!txt) return "";
    var lines = txt.split('\n');
    var indents = lines.map(function (line) {
        var m = line.match(/(^[ \t]*)[^ \t\n]/);
        return m ? m[1] : "";
    });
    indents = indents.filter(function (spc) { return spc > ""; });
    var margin = null;
    indents.forEach(function (indent) {
        if (margin === null) {
            margin = indent;
        } else if (indent.startsWith(margin)) {
            // do nothing
        } else if (margin.startsWith(indent)) {
            margin = indent;
        } else {
            margin = "";
        }
    });
    
    if (margin) {
        txt = txt.replace(new RegExp("^" + margin, "mg"), "");
    }
    return txt;
}


export default {
    // help: jqp.help,
    // lifecycle: require('./dk-lifecycle.js'),
    // counter: require('./dk-counter.js')
    counter,
    text: {
        id2name: id2name,
        cls2id: cls2id,
        count: count_char,
        dedent: dedent
    }
};
