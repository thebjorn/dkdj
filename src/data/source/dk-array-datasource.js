
/*
 *  Use an array of pojos as a data source.
 *
 *  Usage::
 *
 *      const ds = dk.data.ArraySource.create({
 *          data: [
 *              {project: 'Generelt NT', work: '1:03:57'},
 *              {project: 'Tiktok',      work: '2:44:57'},
 *              {project: 'AFR-support', work: '1:06:43'}
 *          ]
 *      });
 */
import {DataSource} from "./dk-datasource-base";
import {sort_by} from "../utils";


export class ArraySource extends DataSource {
    constructor(arr) {
        super();

        if (Array.isArray(arr)) {
            // ArraySource([ {..}, {..}, ..])
            this.data = arr;
        } else {
            // ArraySource({ data: [..]})
            this.data = arr.data;
        }
        // data source items need to have a key
        this.data.map(function (item, i) {
            item.pk = item.pk || i;
        });
    }

    /*
     *  Update the data source with `changes`.
     *
     *  `changes` is a dict with the following shape::
     *
     *      {
     *          <primary-key>: {
     *              <field-name>: {oldval: a, newval: b}
     *              <field-2>: ...
     *          }, ...
     *      }
     *
     *   the data source should unconditionally save changes to all the specified
     *   fields, but only to the records/fields that have changes.
     */
    update(changes) {
        const self = this;
        Object.keys(changes).forEach(function (pk) {
            const change = changes[pk];
            const datarec = self.data[pk];
            for (const attr in change) if (attr !== 'pk' && change.hasOwnProperty(attr)) {
                datarec[attr] = change[attr].newval;
            }
        });
    }

    /*
     *  Return a list of `dk.data.Record`s that satisfy the `request`.
     *
     *  `request` can have the following fields::
     *
     *      start:    the index (0-based) of the first requested record
     *      end:      the index of the last requested record
     *      orphans:  if there are less than this many records after end, include them too
     *      sort:     a list of {field: <fieldname>, direction: "asc"|"desc"
     *      filter:   a json dict containing filter values.
     *
     *  get_records will call callback with the requested records in the following
     *  structure:
     *
     *      recordset   ::= {"fields": field-def, "records": record-data}
     *      field-def   ::= { (fieldname_1: {"name":__, "title":__, "type":__, ...)* }
     *      record-data ::= [ record* ]
     *      record      ::= { (field-name: field-value}* }
     *      field-name  ::= [-_a-zA-Z]+
     *      field-value ::= ...
     *
     *  or structurally:
     *
     *      recordset ::= {
     *          fields: {
     *              fieldname1: {
     *                  name: ___,
     *                  label: ___,
     *                  type: ___,
     *                  ...
     *              }, ...
     *          },
     *          meta: {
     *              start_recnum: __,
     *              totcount: __            // total number of records
     *          },
     *          records: [
     *              {field_name: field_value, ...},
     *              ...
     *          ]
     *      }
     */
    async fetch_records(request) {
        const p = request;
        p.end = (this.data.length - p.orphans < p.end) ? this.data.length : p.end;
        p.start = (p.start > p.end) ? p.end : p.start;

        this.do_sort(p);
        const search_recs = this.do_search(request);
        const result_recs = search_recs.slice(p.start, p.end);

        return new Promise((resolve, reject) => {
            resolve({
                fields: this.get_fields(),
                meta: {
                    totcount: this.data.length,
                    filter_count: search_recs.length,
                    start_recnum: p.start,
                    end_recnum: p.start + result_recs.length - 1
                },
                records: result_recs
            });
        });        
    }
    get_records(request, returns) {
        const p = request;
        p.end = (this.data.length - p.orphans < p.end) ? this.data.length : p.end;
        p.start = (p.start > p.end) ? p.end : p.start;

        this.do_sort(p);
        const search_recs = this.do_search(request);
        const result_recs = search_recs.slice(p.start, p.end);

        returns({
            fields: this.get_fields(),
            meta: {
                totcount: this.data.length,
                filter_count: search_recs.length,
                start_recnum: p.start,
                end_recnum: p.start + result_recs.length - 1
            },
            records: result_recs
        });
    }

    do_search(request) {
        if (!request.search) return this.data;
        return this.data.filter(function (rec) {
            for (let key in rec) if (rec.hasOwnProperty(key)) {
                if (typeof rec[key] === 'string' && rec[key].contains(request.search)) return true;
            }
            return false;
        });
    }

    do_sort(request) {
        if (request.sort && request.sort.length > 0 && request.sort[0].field) {
            this.data = sort_by(this.data, function (rec) {
                return rec[request.sort[0].field];
            });
            if (request.sort[0].direction === 'desc') {
                this.data.reverse();
            }
        }
    }

    /*
     *  Return a list of `dk.data.Field`s representing the "columns" of the
     *  data source (in the correct order).
     */
    get_fields() {
        if (this.data.length > 1) {
            const first = this.data[0];
            const field_defs = {};
            Object.keys(first).forEach(function (key, i) {
                field_defs[key] = {
                    name: key,
                    pos: i,
                    type: typeof first[key]
                };
            });
            return field_defs;
        }
        return {};
    }
}
