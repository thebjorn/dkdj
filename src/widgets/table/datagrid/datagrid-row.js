
import dk from "../../../dk-obj";
import {TableRow} from "../table-row";
import {icon} from "../../dk-icon-library";


/*
 *  A Table row in a data grid.
 */
export class DataGridRow extends TableRow {
    constructor(...args) {
        super({
            // type: 'DataGridRow',
            editing: false,
            command_cell: null,
        }, ...args);
    }

    start_editing(e) {
        e.stopPropagation();
        if (this.editing) return;
        if (!this.table.start_editing(this)) return;  // prevent editing multiple rows
        this.editing = true;
        this.widget().toggle_busy(true);
        this.cells.forEach(cell => {
            cell.start_editing();
            dk.on(cell, 'dirty', (coldef, newval, widget, oldval) => {
                const record = this.record;
                this.table.add_dirty(record, coldef, newval, oldval, widget);
            });
        });
        this.widget('input:first').focus();
        this._save_box_shadow = this.widget().css('box-shadow');
        this.widget().css('box-shadow', '0 0 5px 5px rgba(100,100,100,.2)');
        this.widget().toggle_busy(false);
    }

    stop_editing() {
        this.cells.forEach(function (cell) {
            cell.stop_editing();
        });
        this.widget().css('box-shadow', this._save_box_shadow);
        this.editing = false;
    }

    save_editing() {
        this.cells.forEach(function (cell) {
            cell.save_editing();
        });
        this.table.save();
    }

    undo_editing() {
        this.cells.forEach(function (cell) {
            cell.undo_editing();
        });
    }

    add_command_cell(cell) {
        const self = this;
        this.command_cell = cell;

        const cmd = function (command, color, method) {
            const cmdtxt = (self.table.translate && self.table.translate[command] !== undefined) ? self.table.translate[command] : command;
            const btn = dk.$('<button/>').addClass('btn btn-xs btn-default');
            btn
                .append(icon(command, {cursor: 'pointer', color: color}))
                .append('&nbsp;' + cmdtxt)
                .attr('title', cmdtxt)
                .css({marginLeft: '2px'})
                .click(function (e) {
                    e.preventDefault();
                    self[method]();
                    self.table.stop_editing();
                    return false;
                });

            cell.append(btn);
        };

        if (this.table.commands.includes('save'))   cmd("save", "green", "save_editing");
        if (this.table.commands.includes('remove')) cmd("remove", "#ec520f", "");
        if (this.table.commands.includes('cancel')) cmd("cancel", "#ccc", "undo_editing");

        this.trigger('add-command-cell', cell);
    }

    handlers() {
        this.widget().on('click', (e) => this.start_editing(e));
    }
}
