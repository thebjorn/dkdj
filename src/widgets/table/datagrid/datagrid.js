import dk from "../../../dk-obj";
import {DataTable} from "../data-table";
import {DataGridRow} from "./datagrid-row";
import {dkconsole} from "../../../lifecycle/dkboot/dk-console";

/*
    A datagrid is a data table that can update its data source.
 */
export class DataGrid extends DataTable {
    constructor(...args) {
        super({
            // type: 'datagrid',
            TableRow: DataGridRow,
            commands: ['save', 'remove', 'cancel'],

        }, ...args);
    }

    save() {
        this.table_data.update();     // update datasource
    }

    add_dirty(record, coldef, newval, oldval, widget) {
        this.table_data.page.add_dirty(
            record.pk,
            coldef.field,
            newval,
            oldval,
            widget
        );
    }

    start_editing(row) {
        const self = this;
        if (this.editing) return false;
        this.editing = true;

        dk.$('body')
            .on('keyup.editing', function (e) {
                if (e.keyCode === 27) {
                    row.undo_editing();
                    self.stop_editing();
                }
            })
            .on('click.editing', function () {
                row.undo_editing();
                self.stop_editing();
            });

        this._edit_row = row;
        const cells = this.layout.add_column('append');
        this._command_cells = cells;
        this.widget().addClass('editing');
        cells.forEach(function (cell) {
            cell.css({
                border: 'none !important',
                backgroundColor: 'transparent !important'
            });
        });
        //cells.shift();
        const cell = cells[row.rownum + 1];
        row.add_command_cell(cell);
        return true;  // can edit row
    }

    stop_editing() {
        dk.$('body').off('.editing');
        this._edit_row.stop_editing();
        this._command_cells.forEach(function (cell) {
            cell.remove();
        });
        this._edit_row = null;
        this._command_cells = null;
        this.widget().removeClass('editing');
        this.editing = false;
        this.trigger('stop-editing', this);
        dkconsole.info("Table stop editing", 5);
    }

}
