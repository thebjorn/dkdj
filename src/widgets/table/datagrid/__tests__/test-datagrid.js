import $ from 'jquery';
import page from "../../../../widgetcore/dk-page";
import {DataGrid} from "../datagrid";
import {ArraySource} from "../../../../data/source/dk-array-datasource";
import utidy from "../../../../browser/dk-html";
import {jq_toggle_busy} from "../../../../browser/jquery-plugins";
import dk from "../../../../dk-obj";
import {_reset_counters} from "../../../../core/counter";


test("test-datagrid-render", () => {
    _reset_counters();
    document.body.innerHTML = `
        <div id="work">
            
        </div>
    `;
    const work = $('#work');
    page.initialize(document);

    DataGrid.create_inside(work);

    expect(document.getElementById('work')).toMatchSnapshot();
});


test("test-datagrid-1", () => {
    _reset_counters();
    document.body.innerHTML = `
        <div id="work">
            <div id="mydt-124">
                <table class="table table-bordered table-striped">
                </table>
            </div>
        </div>
    `;
    const work = $('#work');
    page.initialize(document);

    const mydt124 = DataGrid.create_inside('#mydt-124', {

        pagesize: 15,

        datasource: [
            {project: 'FOO-support', work: '1:11:11'},
            {project: 'Generelt NT', work: '2:22:11'},
            {project: 'Tiktok', work: '3:33:11'}
        ],

        columns: {
            work: {
                label: 'Arbeid' //,
                //format: (val) => { return val.replace(/:/g, '.'); }
            },
            project: {
                label: 'Prosjekt',
                format: function (val) {
                    return (val || '').toLowerCase();
                }
            }
        }
    });

    expect(document.getElementById('work')).toMatchSnapshot();
});


test("test-datagrid-2", () => {
    _reset_counters();
    document.body.innerHTML = `
        <div id="work">
            
        </div>
    `;
    const work = $('#work');
    page.initialize(document);
    jq_toggle_busy(dk);

    const mydt125 = DataGrid.create_inside(work, {
        datasource: ArraySource.create({
            data: [ {project: 'Tiktok', work: '3:33:11'} ]
        }),

        columns: {
            work: { label: 'Arbeid' },
            project: { label: 'Prosjekt' }
        }
    });
    work.find('td:first').click();
    console.info('first input', work.find('input:first')); //
    work.find('input:first').val('1:02:03').change();
    console.info("dirtyset:", mydt125.table_data.page.dirtyset);
    expect(Object.keys(mydt125.table_data.page.dirtyset)).toHaveLength(1);
    const dirtyrec = mydt125.table_data.page.dirtyset[0];
    
    // this.assert.equal(dirtyrec.work.oldval, '3:33:11', "model dirty oldval");
    expect(dirtyrec.work.oldval).toEqual('3:33:11');
    expect(dirtyrec.work.newval).toEqual('1:02:03');
    
    //dk.info('data page', mydt125.data.page);

    work.find('i.icon.save').click();
    expect(work.find('td:first').text()).toEqual('1:02:03');
    expect(mydt125.table_data.page.record[0].work).toEqual('1:02:03');
    
});
