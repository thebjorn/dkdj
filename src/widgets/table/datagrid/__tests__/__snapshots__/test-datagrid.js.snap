// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-datagrid-1 1`] = `
<div
  id="work"
>
  
            
  <div
    id="mydt-124"
  >
    
                
    <table
      class="table table-bordered table-striped DataGrid dk-bx"
      id="data-grid-1"
    >
      <thead
        id="dk-bx-1"
      >
        

        <tr
          id="dk-table-1"
        >
          

          <th
            class="string"
            fieldname="work"
            id="dk-table-2"
            scope="col"
            style="cursor: url(//static.datakortet.no/dk/sort-za.cur), pointer;"
            title=""
          >
            <div
              class="SortDirection sort-icon"
              id="sort-direction-1"
              style="float: right; width: 2ex; margin-left: 2ex;"
            />
            Arbeid
          </th>
          

          

          <th
            class="string"
            fieldname="project"
            id="dk-table-3"
            scope="col"
            style="cursor: url(//static.datakortet.no/dk/sort-za.cur), pointer;"
            title=""
          >
            <div
              class="SortDirection sort-icon"
              id="sort-direction-2"
              style="float: right; width: 2ex; margin-left: 2ex;"
            />
            Prosjekt
          </th>
          

        </tr>
        

      </thead>
      
                
      <tfoot
        id="dk-bx-2"
      />
      <tbody
        id="dk-bx-3"
      >
        

        <tr
          class="DataGridRow"
          id="dk-table-4"
          pk="0"
          rownum="0"
        >
          

          <td
            axis="work"
            class="TableCell"
            id="dk-tr-1"
          >
            1:11:11
          </td>
          

          

          <td
            axis="project"
            class="TableCell"
            id="dk-tr-2"
          >
            foo-support
          </td>
          

        </tr>
        

        

        <tr
          class="DataGridRow"
          id="dk-table-5"
          pk="1"
          rownum="1"
        >
          

          <td
            axis="work"
            class="TableCell"
            id="dk-tr-3"
          >
            2:22:11
          </td>
          

          

          <td
            axis="project"
            class="TableCell"
            id="dk-tr-4"
          >
            generelt nt
          </td>
          

        </tr>
        

        

        <tr
          class="DataGridRow"
          id="dk-table-6"
          pk="2"
          rownum="2"
        >
          

          <td
            axis="work"
            class="TableCell"
            id="dk-tr-5"
          >
            3:33:11
          </td>
          

          

          <td
            axis="project"
            class="TableCell"
            id="dk-tr-6"
          >
            tiktok
          </td>
          

        </tr>
        

      </tbody>
    </table>
    
            
  </div>
  
        
</div>
`;

exports[`test-datagrid-render 1`] = `
<div
  id="work"
>
  
            
        
  <table
    class="DataGrid dk-bx busy"
    id="data-grid-1"
  >
    <thead
      id="dk-bx-1"
    >
      

      <tr
        id="dk-table-1"
      />
      

    </thead>
    <tfoot
      id="dk-bx-2"
    />
    <tbody
      id="dk-bx-3"
    />
  </table>
</div>
`;
